### PR TITLE
Use system installed deflate library

### PR DESCRIPTION
**Pull request by bket**
Use pkgconfig if available to locate a libdeflate system library.

If a specific library prefix is given by LIBDEFLATE_PREFIX=somepath, use that.
Otherwise if pkgconfig is installed and the library can be located via that, use that one.
Otherwise (or if enforced by USE_BUNDLED_DEFLATE=yes), fall back to bundled code.

Build tested, with and without deflate library installed, on OpenBSD
amd64 current. Tests pass.

### CODE REVIEW

File: setup.py
Comment (ThomasWaldmann) [inline setup.py]: I guess one would want to emit some warning, so the user will know what's missing / how to fix that.

File: setup.py
Comment (ThomasWaldmann) [inline setup.py]:
```
if libdeflate_found:
```

File: setup.py
Comment (ThomasWaldmann) [inline setup.py]: did you try how that behaves?

i tried this for `borgbackup`, but the problem is that `setup.py` already needs `pkgconfig`  before `setup()` is executed (and could install `pkgconfig` due to `setup_requires`).

File: setup.py
Comment (ThomasWaldmann) [reply setup.py]: it also could emit a info msg when it has found libdeflate via pkg-config and will use that.

File: setup.py
Comment (ThomasWaldmann) [inline setup.py]: what would be the point of `default=True`?

File: setup.py
Comment (ThomasWaldmann) [inline setup.py]: you could add something to force using the bundled code (default: try finding and using the system lib), see how borgbackup deals with the libs in `setup*.py`.

https://github.com/borgbackup/borg/blob/1.2.0/setup.py#L49

**Review by ThomasWaldmann — COMMENTED** [review]
nice addition, i made something similar back then for borg's bundled code / libs.

File: setup.py
Comment (ThomasWaldmann) [reply setup.py]: hmm, if you always fall back to bundled code if pkg-config did not work, it's maybe rather a info msg in all cases.
just tell the user whether it used the pkg-config located system lib or the bundled stuff because ...

Comment (bket) [reply setup.py]: I have a new diff, which mimics the build behavior of borgbackup. This new diff tries to use pkconfig when available. A user can force usage of the bundeld library by having `USE_BUNDLED_DEFLATE` in the environment. Additionally a user can set `LIBDEFLATE_PREFIX` to workaround not having pkgconfig on the system. I will force push this diff in a bit.

File: setup.py
Comment (bket) [reply setup.py]: Yes, you are right.

File: setup.py
Comment (bket) [reply setup.py]: You are right again, this does not work. This has been addressed.

File: setup.py
Comment (bket) [reply setup.py]: I left it there by accident. Removed

File: setup.py
Comment (bket) [reply setup.py]: Done :-)

File: setup.py
Comment (ThomasWaldmann) [inline setup.py]: this is True by default (if the env var is not used).

File: setup.py
Comment (ThomasWaldmann) [inline setup.py]: this is the block executed by default.

there's a problem now if there is no pkgconfig (pc is None) and the user did not set LIBDEFLATE_PREFIX (which is also usually the default).

so, people who do not read the docs and install the lib and pkgconfig before this, will get a build failure - **not a fallback to bundled code**.

**Review by ThomasWaldmann — CHANGES_REQUESTED** [review]
good improvement, but there is a popular failure mode now instead of a fallback.

**Comment by ThomasWaldmann** [comment]
also minor pep8 issue:
```
setup.py:55:21: missing whitespace around operator
```

File: setup.py
Comment (bket) [reply setup.py]: Good catch, should be resolved now.

File: setup.py
Comment (ThomasWaldmann) [inline setup.py]: typo

**Comment by ThomasWaldmann** [comment]
BTW, I edited the top post a bit to reflect the current behaviour.

Could also just be a c&p of the commit comment.

**Comment by ThomasWaldmann** [comment]
BTW, our tests of course only test the default behaviour and do not set any of these env vars yet - but I suspect you tried all modes manually?

File: setup.py
Comment (ThomasWaldmann) [reply setup.py]: same typo also in commit comment.

**Comment by bket** [comment]
Fixed :-)

File: setup.py
Comment (bket) [reply setup.py]: Fixed

**Comment by bket** [comment]
> BTW, our tests of course only test the default behaviour and do not set any of these env vars yet - but I suspect you tried all modes manually?

Yes, tried all modes manually. So far, behavioud is as expected.

**Review by ThomasWaldmann — APPROVED** [review]
LGTM!

**Comment by ThomasWaldmann** [comment]
@dcwatson I think this is a good and very flexible improvement to `deflate` and it especially enables dist package maintainers to not use the bundled code (or even remove it) if they rather prefer to use the `libdeflate` library from another dist package.

Avoids duplication and eases maintenance, especially in case of critical bug or security fixes, they then only need to get fixed at one place.

Merge?

File: setup.py
Comment (dcwatson) [inline setup.py]: `USE_BUNDLED_DEFLATE=0` will use the bundled libdeflate, which seems counterintuitive, especially if something is trying to set that env to a true or false value based on a condition (rather than setting/not setting). I think having a list of true or false values and some documentation will take care of this.

File: setup.py
Comment (dcwatson) [inline setup.py]: Do we need this warning if using the bundled version?

File: setup.py
Comment (ThomasWaldmann) [reply setup.py]: that's basically my quick and dirty code @bket took from borgbackup.

we only support not setting it or setting it to =yes (or in fact =anything, which all evals to true).

i agree that true / false / True / False / yes / no / 0 / 1 (anything else?) would be cleaner.

File: setup.py
Comment (ThomasWaldmann) [reply setup.py]: No, when forcing the bundled code, no pkgconfig is used.

**Comment by ThomasWaldmann** [comment]
@bket If you don't mind, I'll fix it now according to the review comments of @dcwatson ...

File: setup.py
Comment (ThomasWaldmann) [reply setup.py]: no warning any more if not used anyway.

File: setup.py
Comment (ThomasWaldmann) [reply setup.py]: made it more clean/flexible now.

**Comment by bket** [comment]
> @bket If you don't mind, I'll fix it now according to the review comments of @dcwatson ...

@ThomasWaldmann, thank you for taking care of this! Changes lgtm, and make sense.

**Comment by ThomasWaldmann** [comment]
@dcwatson have a look now!

**Comment by ThomasWaldmann** [comment]
@bket if this gets merged and released, it will unblock some borg PRs, that's why I am trying to speed this up a bit. :-)

**Comment by ThomasWaldmann** [comment]
maybe squash all commits together instead of merging all the separate commits.

**Comment by ThomasWaldmann** [comment]
trying pytest within cibuildwheel now vvv.

**Comment by ThomasWaldmann** [comment]
yay! all except pypy tested OK.

Guess pypy breakage is not new and unrelated to this PR?
```
  ImportError while importing test module '/project/tests/test_compress.py'.
  Hint: make sure your test modules/packages have valid Python names.
  Traceback:
  /opt/python/pp37-pypy37_pp73/lib-python/3/importlib/__init__.py:127: in import_module
      return _bootstrap._gcd_import(name[level:], package, level)
  /project/tests/test_compress.py:6: in <module>
      import deflate
  E   ImportError: /tmp/tmp.Dh9bFArfRD/venv/site-packages/deflate.pypy37-pp73-x86_64-linux-gnu.so: undefined symbol: Py_Initialize
```

**Comment by ThomasWaldmann** [comment]
@dcwatson have a look! before merging i need to remove that temporary commit.

https://github.com/dcwatson/deflate/runs/5489155309?check_suite_focus=true

**Comment by ThomasWaldmann** [comment]
@dcwatson considering this works and wheels are even tested now, i'ld like to merge this if you don't object.

this will remove the pypy wheel build (which likely was broken before also, just nobody noticed).

i'll remove that temporary commit now, which was just for demonstration purposes, but we still have the test results linked above.

**Comment by ThomasWaldmann** [comment]
temporary commit removed, rebased on current master.

**Comment by ThomasWaldmann** [comment]
As there were no more objections, I merged this now.

In case we find more to do / to fix, it can be done in a later PR.

Next thing we need is a new release. @dcwatson can you do that or give me permissions on pypi.org to do that? #26

**Pull request merged**
